lib/THC/THCReduce.cuh

```diff
@@ -0,0 +1,319 @@
+                                ModifyOp modifyOp,
+                                ReduceOp reduceOp) {
+  const IndexType sliceIndex = getReduceNoncontigDimSliceIndex<IndexType>();
+
+  if (sliceIndex >= totalSlices) {
+    return;
+  }
+
+  // Each thread picks a point in `out` and `in` for which it is
+  // producing the reduction
+  const IndexType outOffset =
+    IndexToOffset<IndexType, ADims>::get(sliceIndex, out);
+  const IndexType inBaseOffset =
+    IndexToOffset<IndexType, BDims>::get(sliceIndex, in);
+
+  // For each point in reductionSize, reduce into `r`
+  IndexType inOffset = inBaseOffset;
+  float r = init;
+
+  for (IndexType i = 0; i < reductionSize; ++i) {
+    r = reduceOp(r, modifyOp(in.data[inOffset]));
+    inOffset += reductionStride;
+  }
+
+  // Write out reduced value
+  out.data[outOffset] = r;
+}
+
+template <typename IndexType>
+__device__ __forceinline__ IndexType getReduceContigDimSliceIndex() {
+  // Each block handles one slice
+  return getLinearBlockId<IndexType>();
+}
+
+// Kernel that handles an entire reduction of a slice of a tensor per
+// each block
+template <typename ModifyOp, typename ReduceOp, typename IndexType, int ADims, int BDims>
+__global__ void
+THCudaTensor_reduceContigDim(TensorInfo<IndexType> out,
+                             TensorInfo<IndexType> in,
+                             IndexType reductionSize,
+                             IndexType totalSlices,
+                             float init,
+                             ModifyOp modifyOp,
+                             ReduceOp reduceOp) {
+  const IndexType sliceIndex = getReduceContigDimSliceIndex<IndexType>();
+
+  if (sliceIndex >= totalSlices) {
+    return;
+  }
+
+  // Get the offset in `out` for the reduction
+  const IndexType outOffset =
+    IndexToOffset<IndexType, ADims>::get(sliceIndex, out);
+
+  // Get the base offset in `in` for this block's reduction
+  const IndexType inBaseOffset =
+    IndexToOffset<IndexType, BDims>::get(sliceIndex, in);
+
+  // Each thread in the block will reduce some subset of elements in
+  // the slice. The elements are guaranteed contiguous starting at
+  // `inBaseOffset`.
+  float r = init;
+  for (IndexType i = threadIdx.x; i < reductionSize; i += blockDim.x) {
+    r = reduceOp(r, modifyOp(in.data[inBaseOffset + i]));
+  }
+
+  // Reduce within the block
+  extern __shared__ float smem[];
+  smem[threadIdx.x] = r;
+
+  // First warp will perform reductions across warps
+  __syncthreads();
+  if ((threadIdx.x / 32) == 0) {
+    r = init;
+    for (int i = 0; i < blockDim.x; i += 32) {
+      r = reduceOp(r, modifyOp(smem[i + threadIdx.x]));
+    }
+
+    // Each lane participating writes out a value
+    smem[threadIdx.x] = r;
+  }
+
+  // First thread will perform reductions across the block
+  __syncthreads();
+  if (threadIdx.x == 0) {
+    r = init;
+#pragma unroll
+    for (int i = 0; i < 32; ++i) {
+      r = reduceOp(r, modifyOp(smem[i]));
+    }
```

Comment:
Cool, that's impressive! What's the difference between the first and second column here?

And the code would still be correct if the warp size wasn't 32, right? We only rely on it for performance, not correctness, if I read the code correctly.